### PR TITLE
Fix add_roots_for()

### PR DESCRIPTION
**Pull request by neithernut**
During investigation of issue #601, @matthiasbeyer found that, apparently, the function `add_roots_for()` misbehaves. Thi PR is aimed at fixing that function.

Note that this PR contains a rewrite of that function. It appears that that rewrite either breaks some functionality or uncovers another bug. I'd suggest investigating whether the other two fixes resolve the issue, e.g. whether the rewrite can be omitted. That would be a job for @matthiasbeyer as he found the original bug.

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Shall I wait until this actually builds?

**Comment by matthiasbeyer** [comment]
Now it segfaults in `src/compositor/buffer/buffer.c`, [here](https://github.com/waysome/waysome/blob/develop/src/compositor/buffer/buffer.c#L175) as the ptr is not valid. 

I'm not sure, though, whether this is already implemented. Assuming from the ptr value, it is not (I'm too lazy to look it up right now).

So it _seems_ that your patches fix it. I will investigate further, though.

**Comment by matthiasbeyer** [comment]
I tried the following configurations:
- shutdown
- shutdown, terminal
- terminal
- terminal, shutdown

They all worked both for shutdown and starting the terminal (while segfaulting because of the false ptr value mentioned above). So I think your patches are sufficient:

**ACK**

This is for the HEAD of the PR! I would like to re-review after you squashed, to be sure everything still works.

**Comment by neithernut** [comment]
**Read the comment I wrote to this PR!**
I **explicitely stated** that the current state introduces new problems but you should test it with **the first two commits only!**

**Comment by neithernut** [comment]
I found a minor bug in `ws_hotkey_dag_remove()`. Apparently, the event is removed from the tree, but not all the nodes that could be removed are, in fact, removed.
I also found the reason why this minor bug didn't show up previously: the insertion method is imperfect. It adds one unnecessary level in the dag node table.
I'm working on fixing those issues.

**Comment by neithernut** [comment]
_Slowly_ getting closer here.

**Comment by neithernut** [comment]
_Now_ this is ready for review and merge (_finally_).

**Comment by matthiasbeyer** [comment]
Bad news:

```
#0  0x000000000040e71d in destruct_tab_node (tab_node=0x4, depth=12 '\f') at /home/m/dev/c/waysome/src/input/hotkey_dag.c:404
#1  0x000000000040e72a in destruct_tab_node (tab_node=0x67da30, depth=13 '\r') at /home/m/dev/c/waysome/src/input/hotkey_dag.c:404
#2  0x000000000040e72a in destruct_tab_node (tab_node=0x6735b0, depth=14 '\016') at /home/m/dev/c/waysome/src/input/hotkey_dag.c:404
#3  0x000000000040e72a in destruct_tab_node (tab_node=0x672100, depth=15 '\017') at /home/m/dev/c/waysome/src/input/hotkey_dag.c:404
#4  0x000000000040e72a in destruct_tab_node (tab_node=0x67cee0, depth=16 '\020') at /home/m/dev/c/waysome/src/input/hotkey_dag.c:404
#5  0x000000000040e090 in ws_hotkey_dag_deinit (entry_node=0x629800 <ws_hotkeys_ctx>) at /home/m/dev/c/waysome/src/input/hotkey_dag.c:144
#6  0x000000000040da6e in hotkeys_deinit (dummy=0x0) at /home/m/dev/c/waysome/src/input/hotkeys.c:264
#7  0x000000000041d8ea in ws_cleaner_run () at /home/m/dev/c/waysome/src/util/cleaner.c:99
#8  0x000000000040adf3 in main (argc=1, argv=0x7fffffffe118) at /home/m/dev/c/waysome/src/main.c:177
```

Steps to reproduce:

```
$ gdb ./src/waysome
run
<hit STRG-ALT-Q>
```

**Comment by neithernut** [comment]
Damn. Well, at least this is easily testable, but strangely, deinitialization _should_ be covered by the tests already, _or is it_?

Note: `depth` _must_ be `<= 4`.

**Comment by neithernut** [comment]
Turns out deinitialization was indeed, completely left out when writing the test cases.

**Comment by neithernut** [comment]
The problem described by @matthiasbeyer was fixed and he also kindly provided updates to the tests. Can I get an ACK?

---

WTF, _now_ the tests start failing.

**Comment by neithernut** [comment]
Some minor fix seemed to have fixed the problem, please test thorougly.

**Comment by matthiasbeyer** [comment]
I did a short test and everything seems to work as expected now. But, I think I'm too tired right now to do a in-depth test and therefor I'd like to postpone this until tomorow?

**Comment by neithernut** [comment]
I'm OK with this, but I'd like to get this merged tomorrow as the bug(s) fixed by this PR are quite nasty.

File: src/input/hotkey_dag.c [hunk 1]
```diff
@@ -403,34 +388,24 @@ create_tab_node(void) {
...
     // iterate over all the nodes
     while (cur_node-- > (void**) tab_node) {
         // destruct children
-        if (depth) {
-            destruct_tab_node(*cur_node, depth);
-        } else {
-            destruct_dag_node((struct ws_hotkey_dag_node*) cur_node);
+        if (depth > 0) {
+            destruct_tab_node(*cur_node, depth - 1);
+        } if (*cur_node) {
```
Comment (matthiasbeyer) [inline src/input/hotkey_dag.c]: Is this supposed to be an `if` ? If yes, style fuckup, if not bug!

**Comment by matthiasbeyer** [comment]
ACK from me. Let @TheNeikos check it, though!

**Comment by neithernut** [comment]
@TheNeikos ACK'ed verbally.

**Pull request merged**